working/0093 - Gradual Language Change Migration/0094 - Per Library Language Version Selection/strawman.md

```diff
@@ -78,14 +118,28 @@ The `pub` tool configures this default language level for a package based on its
 
-Unpackaged libraries (libraries with `file:` or `http:` URIs, anything except `package:` and `dart:`) cannot use this approach to get a different default. Unpackaged libraries include *tests* in pub packages. Tools will get a `--default-package` flag that allows users to set a package that all unpackaged libraries are considered as belonging to. This should be used when running tests or other pub-package related files that are not in the Dart package. (If Dart ever gets a notion of package-privacy, this feature will allow tests to pierce the privacy of their own package, without providing a general way for code to do so).
+Unpackaged libraries (libraries with `file:` or `http:` URIs, anything except `package:` and `dart:`) cannot use this approach to get a different default. Unpackaged libraries include *tests* in pub packages.
+
+**Bob: Tests *are* part of the containing package as far as the rest of the
+system is concerned. They cannot be *imported* from other packages, but they are
+definitely part of the package. We can just put the default language version in
+the .packages file and use that for libraries not under lib. Pub can add that
+line based on the SDK constraint.**
+
+Tools will get a `--default-package` flag that allows users to set a package that all unpackaged libraries are considered as belonging to. This should be used when running tests or other pub-package related files that are not in the Dart package. (If Dart ever gets a notion of package-privacy, this feature will allow tests to pierce the privacy of their own package, without providing a general way for code to do so).
 Alternatively, unpackaged libraries can have their version specified directly on the command line for the compiler (`dart --default-version=2.2`). With no flags, neither `--default-package` or `--default-version`, unpackaged libraries default to the newest language version.
 
 
 (If we allow un-packaged code to act as if it was in a package, e.g., for testing, it will inherit the package's language level, which is reasonable since that code is usually in the same pub package, just not in the exported Dart package).
 
 It *is* an option to only support per-package versioning, and not have a per library versioning option. It may increase the migration cost for a package, which encourages putting it off until the latest possible moment, but it may also encourage *completing* a migration sooner instead of leaving a few less-important files hanging.
 
+**Bob: One concern is that package authors are disincentivized from migrating.
+To migrate, they logically must raise the minimum SDK constraint, but that in
+turn makes it impossible for users on older SDKs to use their package.
+Hopefully, the new features are delightful enough to mitigate this, but we
+should keep it in mind.**
```

Comment:
If package P1 (which allows for language version levels k1..k2) imports something from package P2 (which allows for language version levels m1..m2) then it seems reasonable to use language level n which is some number in the intersection of k1..k2 and m1..m2 (as opposed to insisting on level k1 and then failing if k1 isn't in the range m1..m2).

Is it even likely that we would be able to allow for any kind of version mix without going into version resolution of a similar kind as that of pub?